### PR TITLE
E1559. Refactoring JoinTeamRequestsController and InvitationController

### PR DESCRIPTION
**Pull request by mandgerohit**
Completed all the required functionality.

### CODE REVIEW

File: app/mailers/mailer.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 class Mailer < ActionMailer::Base
 
   if Rails.env.development? || Rails.env.test?
-    default from: 'expertiza.development@gmail.com'
+    default from: 'noreply@expertiza.com'
```
Comment (rkyadav-ncsu) [inline app/mailers/mailer.rb]: please revert back this to original defaults.

File: app/mailers/mailer.rb [hunk 2]
```diff
@@ -14,9 +14,10 @@ def generic_message(defn)
     @new_pct = defn[:body][:new_pct]
     @avg_pct = defn[:body][:avg_pct]
     @assignment = defn[:body][:assignment]
+    @link = defn[:body][:link]
 
     if Rails.env.development? || Rails.env.test?
-      defn[:to] = 'expertiza.development@gmail.com'
+      defn[:to] = 'rshah5@ncsu.edu'
```
Comment (rkyadav-ncsu) [inline app/mailers/mailer.rb]: revert back this to original default.

File: app/mailers/mailer.rb (hunk 1 above)
Comment (mandgerohit) [inline app/mailers/mailer.rb]: Done changes.

File: app/mailers/mailer.rb (hunk 2 above)
Comment (mandgerohit) [inline app/mailers/mailer.rb]: Done changes

File: app/controllers/student_teams_controller.rb [hunk 3]
```diff
@@ -51,8 +51,7 @@ def view
         break
       end
     end
-
-    @teammate_review_allowed = true if @student.assignment.find_current_stage=='Finished' || @current_due_date&&(@current_due_date.teammate_review_allowed_id ==3 ||@current_due_date.teammate_review_allowed_id ==2) #late(2) or yes(3)
+    @teammate_review_allowed = true if @current_due_date&&(@current_due_date.teammate_review_allowed_id ==3 ||@current_due_date.teammate_review_allowed_id ==2) #late(2) or yes(3)
```
Comment (yangsong8) [inline app/controllers/student_teams_controller.rb]: Please take a look on "find_current_stage" in assignment.rb. If all the deadlines have passes, this method will return a string "Finished". We made the system allow teammate review even after the last round of deadline.

File: app/controllers/join_team_requests_controller.rb [hunk 4]
```diff
@@ -3,40 +3,38 @@ class JoinTeamRequestsController < ApplicationController
...
+    ['Student', 'Instructor', 'Teaching Assistant'].include?(current_role_name) #people with this roles can only access the function provied by the controller 
   end
 
-  def index
-    @join_team_requests = JoinTeamRequest.all
-
-    respond_to do |format|
-      format.html # index.html.erb
-      format.xml  { render :xml => @join_team_requests }
+  private
```
Comment (yangsong8) [inline app/controllers/join_team_requests_controller.rb]: By convention the all the methods below "private" should be private method.

File: app/controllers/invitation_controller.rb [hunk 5]
```diff
@@ -1,128 +0,0 @@
-class InvitationController < ApplicationController
```
Comment (yangsong8) [inline app/controllers/invitation_controller.rb]: This renaming was done following convention.

File: app/controllers/join_team_requests_controller.rb [hunk 6]
```diff
@@ -3,40 +3,38 @@ class JoinTeamRequestsController < ApplicationController
...
-      format.xml  { render :xml => @join_team_requests }
+  private 
+  def render_request
+      respond_to do |format|
+      format.html 
+      format.xml  { render :xml => @join_team_request } #displays the join team instance
     end
   end
 
+  def index
```
Comment (yangsong8) [inline app/controllers/join_team_requests_controller.rb]: Do we have a index page to for join_team_requests? Why do a JoinTeamRequest.all?

File: app/controllers/student_teams_controller.rb (hunk 3 above)
Comment (mandgerohit) [inline app/controllers/student_teams_controller.rb]: I am not sure who is the owner of this change. In OSS project, we changed only Invitations and JoinTeamRequest controllers. These changes were done by some other team and their changes must have been merged before us.

File: app/controllers/join_team_requests_controller.rb (hunk 4 above)
Comment (mandgerohit) [inline app/controllers/join_team_requests_controller.rb]: render_request is already declared private in Git. Can you please provide the method names?

File: app/controllers/join_team_requests_controller.rb (hunk 6 above)
Comment (mandgerohit) [inline app/controllers/join_team_requests_controller.rb]: The index method was already present. We just had to map the redundant code into another private method. That was the task assigned to us. Although we didn't have the index page for join_team_requests, we didn't delete the method as we didn't know the overall impact on the system.

**Comment by yangsong8** [comment]
When I impersonate a student and tried to invite a student which is not in the current assignment. I got a green message saying "*** is not a participant of this assignment". However the invitation was displayed under "Sent Invitations":

image

Some wired output from received request list:
image

When team member decline a join team request, I got the error message below:
image

**Comment by yangsong8** [comment]
By the way, team, thank you for your work. For the issue I mentioned above, you do not need to fix them. Good luck with your finals.

**Pull request closed without merging**